Implement fixed-overhead polling; add `wait_until_done()` for Transactions

NHDaly opened the pull request:
- Implement the fixed (~1%~ ~5%~ 10%) overhead polling algorithm.
- Add a user-facing `wait_until_done(txn)` function.
- Implement `exec()` in terms of `exec_async()` + `wait_until_done()`.
- Update tests

This implements the polling algorithm we discussed at Monday's API meeting:
https://github.com/RelationalAI/rai-sdk-issues/issues/58
> One way to approach this is "wait 1% of the time the transaction has been running so far", then exec() adds at most 1% to the duration of the transaction. With the above backoff if I'm reading it correctly you double the duration of transactions that take slightly more than 0.5 seconds, and it falls off gradually e.g. add 20% to transactions that take 5 seconds.

NHDaly commented:
If we find that 1% overhead is polling too frequently, we can always bump it up to 5% or something like that.

bachdavi commented on src/api.jl:
```diff
@@ -50,33 +50,70 @@ function Base.show(io::IO, e::HTTPError)
-
-    for duration in backoff
+    @assert overhead_rate >= 0.0
+    delay_rate = 1.0 + overhead_rate
```
Should this not be `*`. In the current implementation, it seems we multiply the `delay_rate` with the current `duration` of the transaction. This means we sleep for the current duration of the transaction plus the overhead rate, and not just the overhead rate. Or am I missing something :)

NHDaly replied:
Good call! 😅 sorry and thanks! I did it right in my Rel model 😅

bachdavi reviewed (commented):
Looking good 🙂
Only one two questions:  
- I'm not sure if we are actually adding only `overhead` to the transaction in the current implementation, but its also early morning ^^
- `1%` seems pretty aggressive, especially for the integration tests there we will quickly reach the `14` iterations...

tveldhui commented on src/api.jl:
```diff
@@ -50,33 +50,70 @@ function Base.show(io::IO, e::HTTPError)
             return nothing
         end
+        current_delay = time_ns() - start_time
+        duration = (current_delay * delay_rate) / 1e9
```
As David said, I think this should just be `duration = overhead_rate * current_delay / 1e9`

NHDaly replied:
hehe woops, yes, i think you are both correct. Thank you! Will fix this now.

tveldhui commented on src/api.jl:
```diff
@@ -50,33 +50,70 @@ function Base.show(io::IO, e::HTTPError)
-    for duration in backoff
+    @assert overhead_rate >= 0.0
+    delay_rate = 1.0 + overhead_rate
+    start_time = time_ns()
```
This should be from the moment the transaction request is received (is there a timestamp we can access)?

tveldhui replied:
Oh I guess this is client-side so it doesn't really matter?

NHDaly replied:
Yeah, exactly. The client is _creating_ the transaction, so that start time, _really is_ the start time. 👍

NHDaly replied:
ahhhh but no you're right, this was _not_ including the time since we first created the TXN in the calculation for how long to poll for.
You're right! Okay let me see if i can work that in.
If not, the `txn` object that was returned from the server actually _does_ include the start time from the server's perspective, so in the worst case we can use that. Good catch, @tveldhui! Thanks

NHDaly commented:
Thanks all, i've addressed your feedback. Here are some screen recordings showing the behavior after all of that has been incorporated:

I think this seems pretty reasonable. This is 5% overhead + including the initial create transaction time.
For a 2 second txn, we create then poll 4 times. For a 15 second txn, we poll 32 times. That seems reasonable to me. Does that look good to everyone?


https://user-images.githubusercontent.com/1582097/185000980-b7dcf890-ee45-4c45-8625-a66c9c9a9122.mov


https://user-images.githubusercontent.com/1582097/185000985-650410d0-37b0-4543-9fec-68633e75f654.mov

NHDaly commented on test/integration.jl:
```diff
@@ -1,14 +1,14 @@
 # -----------------------------------
 # context & setup
 
 # This will poll for a maximum of ~600 seconds in 14 steps.
```
i'm not sure... 🤔

NHDaly replied:
haha no, wait, it doesn't hold at all. good call.

NHDaly replied:
adjusted, thanks

NHDaly commented on src/api.jl:
```diff
@@ -50,33 +50,70 @@ function Base.show(io::IO, e::HTTPError)
-# reached. Polling frequency is controlled by an `ExponentialBackOff`. If `throw_on_max_n`
-# is set to true, this will throw if the maximum number of iterations are reached.
-function _poll_until(
+# reached. Polling frequency is controlled to minimize overhead, by keeping the overhead
+# from sleeping within a specified overhead rate. If `throw_on_max_n` is set to true, this
```
@bachdavi: Hopefully this is somewhat clearer?:
> # Polling frequency is computed to minimize overhead: we carefully set the sleep time
> # between polls to a fraction of the time waited so far, so that after any given sleep, we
> # cannot have missed the upstream result by more than `overhead_rate` of the _actual time_.

NHDaly commented:
Okay this is ready for a final review. Please take another look everyone! Thank you!

NHDaly commented:
I've bumped to 10% overhead, per discussion in the SDK meeting. We agreed that we will start conservative, and then become more aggressive if we find no issues from the increased load to the Infrastructure.

We are also going to increase the fast-path timeout from 1 second to 2 seconds, to reduce the number of requests to Infrastructure, and we can drop it back down if we find that things are okay.

NRHelmi reviewed (approved):
Thanks @NHDaly

NHDaly commented:
Thank you Helmi!
